lib/registry-secrets.ts

```
@@ -0,0 +1,95 @@
+ * password entries keyed by hostname:port registry addresses).
+ */
+export class RegistrySecretValidator {
+	private registrySecretJsonSchema = {
+		// Sample valid registrySecrets JSON contents:
+		//   {  "docker.example.com": {"username": "ann", "password": "hunter2"},
+		//      "https://idx.docker.io/v1/": {"username": "mck", "password": "cze14"}
+		//   }
+		type: 'object',
+		patternProperties: {
+			'^\\S+$': {
+				type: 'object',
+				properties: {
+					username: { type: 'string' },
+					password: { type: 'string' },
+				},
+				additionalProperties: false,
+			},
+		},
+		additionalProperties: false,
+	};
+	private validator: ajv.Ajv;
+	private validateFunction: ajv.ValidateFunction;
+
+	constructor() {
+		this.validator = new ajv();
+		this.validateFunction = this.validator.compile(
+			this.registrySecretJsonSchema,
+		);
+	}
+
+	/**
+	 * Validate the given JSON object against the registry secrets schema.
+	 * Throw an error if validation fails.
+	 * @param parsedJson The result of calling JSON.parse()
+	 * @returns The input object cast to the RegistrySecrets type if validation succeeds
+	 * @throws Throws an error if validation fails
+	 */
+	public validateRegistrySecrets(parsedJson: object): RegistrySecrets {
+		const valid = this.validateFunction(parsedJson);
+		if (!valid) {
+			throw new RegistrySecretValidationError(
+				this.validator.errorsText(this.validateFunction.errors));
+		}
+		return parsedJson as RegistrySecrets;
+	}
+
+	/**
+	 * Call JSON.parse() on the given string, then validate the result against
+	 * the registry secrets schema.
+	 * @param json String containing a JSON representation of registry secrets
+	 * @returns A JS object that complies with the RegistrySecrets interface
+	 * @throws Throws an error if parsing or validation fails
+	 */
+	public parseRegistrySecrets(json: string): RegistrySecrets {
+		return this.validateRegistrySecrets(JSON.parse(json));
```

It is possible for `JSON.parse` to throw - is it ok to just let that bubble up here or should it be accounted for?

---

Yes, parsing or validation failure are "notified" by throwing an error. There is a `@throws` JSDoc line above. Users of this method are expected to try/catch it.